pkg/gatewaysecret/handler.go

```diff
@@ -0,0 +1,182 @@
+		if caCert.Status.NotBefore != nil && gwSecretLastModifiedAt.After(caCert.Status.NotBefore.Time) {
+			return nil
+		}
+	}
+
+	gwSecret.Data["tls.crt"] = rootSecret.Data["tls.crt"]
+	gwSecret.Data["tls.key"] = rootSecret.Data["tls.key"]
+	gwSecret.Data["ca.crt"] = rootSecret.Data["ca.crt"]
+	return gsh.update(ctx, gwSecret)
+}
+
+func (gsh *GatewaySecretHandler) newGatewaySecret(rootSecret *apicorev1.Secret) *apicorev1.Secret {
+	gwSecret := &apicorev1.Secret{
+		TypeMeta: apimetav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: apicorev1.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: apimetav1.ObjectMeta{
+			Name:      gatewaySecretName,
+			Namespace: istioNamespace,
+		},
+		Data: map[string][]byte{
+			"tls.crt": rootSecret.Data["tls.crt"],
+			"tls.key": rootSecret.Data["tls.key"],
+			"ca.crt":  rootSecret.Data["ca.crt"],
+		},
+	}
+	return gwSecret
+}
+
+func (gsh *GatewaySecretHandler) findGatewaySecret(ctx context.Context) (*apicorev1.Secret, error) {
+	return GetGatewaySecret(ctx, gsh.kcpClient)
+}
+
+func (gsh *GatewaySecretHandler) create(ctx context.Context, secret *apicorev1.Secret) error {
+	gsh.updateLastModifiedAt(secret)
+	if err := gsh.kcpClient.Create(ctx, secret); err != nil {
+		return fmt.Errorf("failed to create secret %s: %w", secret.Name, err)
+	}
+	return nil
+}
+
+func (gsh *GatewaySecretHandler) update(ctx context.Context, secret *apicorev1.Secret) error {
+	gsh.updateLastModifiedAt(secret)
+	if err := gsh.kcpClient.Update(ctx, secret); err != nil {
+		return fmt.Errorf("failed to update secret %s: %w", secret.Name, err)
+	}
+	return nil
+}
+
+func (gsh *GatewaySecretHandler) updateLastModifiedAt(secret *apicorev1.Secret) {
+	if secret.Annotations == nil {
+		secret.Annotations = make(map[string]string)
+	}
+	secret.Annotations[LastModifiedAtAnnotation] = apimetav1.Now().Format(time.RFC3339)
+}
+
+func GetGatewaySecret(ctx context.Context, clnt client.Client) (*apicorev1.Secret, error) {
+	secret := &apicorev1.Secret{}
+	if err := clnt.Get(ctx, client.ObjectKey{
+		Name:      gatewaySecretName,
+		Namespace: istioNamespace,
+	}, secret); err != nil {
+		return nil, fmt.Errorf("failed to get secret %s: %w", gatewaySecretName, err)
+	}
+	return secret, nil
+}
+
+func GetValidLastModifiedAt(secret *apicorev1.Secret) (time.Time, error) {
+	if gwSecretLastModifiedAtValue, ok := secret.Annotations[LastModifiedAtAnnotation]; ok {
+		if gwSecretLastModifiedAt, err := time.Parse(time.RFC3339, gwSecretLastModifiedAtValue); err == nil {
+			return gwSecretLastModifiedAt, nil
+		}
+	}
+	return time.Time{}, errCouldNotGetLastModifiedAt
+}
+
+func WatchChangesOnRootCertificate(clientset *kubernetes.Clientset, gatewaySecretHandler *GatewaySecretHandler,
+	log logr.Logger,
+) {
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+
+	secretWatch, err := clientset.CoreV1().Secrets(istioNamespace).Watch(ctx, apimetav1.ListOptions{
```

Comment:
In this point a strong compile-time dependency on k8s library code is introduced.
It makes it impossible to properly unit-test our logic contained in the `for event := ...` loop.